core/src/main/java/org/apache/accumulo/core/file/blockfile/cache/lru/CachedBlockQueue.java

```diff
@@ -17,98 +17,89 @@
-      initialSize++;
-    queue = new PriorityQueue<>(initialSize);
-    heapSize = 0;
+  public CachedBlockQueue(final long maxSize, final long blockSize) {
+    Preconditions.checkArgument(maxSize > 0L);
+    Preconditions.checkArgument(blockSize > 0L);
+
+    final long initialSize = (maxSize + blockSize - 1L) / blockSize;
+
+    this.queue = new PriorityQueue<>(Math.toIntExact(initialSize));
+    this.heapSize = 0L;
     this.maxSize = maxSize;
   }
 
   /**
    * Attempt to add the specified cached block to this queue.
    *
-   * <p>
-   * If the queue is smaller than the max size, or if the specified element is ordered before the
-   * smallest element in the queue, the element will be added to the queue. Otherwise, there is no
-   * side effect of this call.
+   * If the block fits within the max heap size, the element will be added to
+   * the queue. Otherwise, if the block being added is sorted larger than the
+   * smallest element, the block is inserted and the smallest element is removed
+   * from the queue.
    *
-   * @param cb
-   *          block to try to add to the queue
+   * @param cb block to try to add to the queue
+   * @return true if the CachedBlock was added to the queue
    */
-  public void add(CachedBlock cb) {
-    if (heapSize < maxSize) {
-      queue.add(cb);
+  public boolean add(final CachedBlock cb) {
+    if (available() > 0) {
       heapSize += cb.heapSize();
-    } else {
-      CachedBlock head = queue.peek();
-      if (cb.compareTo(head) > 0) {
-        heapSize += cb.heapSize();
-        heapSize -= head.heapSize();
-        if (heapSize > maxSize) {
-          queue.poll();
-        } else {
-          heapSize += head.heapSize();
-        }
-        queue.add(cb);
-      }
+      return queue.add(cb);
     }
-  }
-
-  /**
-   * Get a sorted List of all elements in this queue, in descending order.
-   *
-   * @return list of cached elements in descending order
-   */
-  public CachedBlock[] get() {
-    LinkedList<CachedBlock> blocks = new LinkedList<>();
-    while (!queue.isEmpty()) {
-      blocks.addFirst(queue.poll());
+    final CachedBlock mruPeek = queue.peek();
+    if (mruPeek.compareTo(cb) > 0) {
+      return false;
     }
-    return blocks.toArray(new CachedBlock[blocks.size()]);
+    heapSize += cb.heapSize();
+    heapSize -= mruPeek.heapSize();
+    queue.poll();
+    return queue.add(cb);
   }
 
   /**
-   * Get a sorted List of all elements in this queue, in descending order.
+   * Get a copy of all elements in this queue, in LRU ascending order.
    *
-   * @return list of cached elements in descending order
+   * @return Collection of cached elements in queue order
    */
-  public LinkedList<CachedBlock> getList() {
-    LinkedList<CachedBlock> blocks = new LinkedList<>();
-    while (!queue.isEmpty()) {
-      blocks.addFirst(queue.poll());
-    }
-    return blocks;
+  public Collection<CachedBlock> getBlocks() {
+    CachedBlock[] blocks = queue.toArray(new CachedBlock[0]);
```

Comment:
Could you simplify this ( may also result in one less object translation:
`
 Collections.sort(new ArrayList<>(queue), Collections.reverseOrder());
`

Reply:
@phrocker I attempted something like that already, but this ended up being the most simple way.  In your proposal the `Collections.sort` routine does not return a value, so the ArrayList needs to be created outside the list and then returned, much as I did with the native array.